covenant-signer/keyutils/bip32deriv.go

```diff
@@ -0,0 +1,129 @@
+package keyutils
+
+import (
+	"encoding/hex"
+	"fmt"
+	"strconv"
+	"strings"
+
+	"github.com/btcsuite/btcd/btcutil/hdkeychain"
+)
+
+const (
+	HardenedPostfix         = "h"
+	ExpectedDerivationDepth = 5
+)
+
+func parseHardened(elem string) (uint32, error) {
+	// valid hardened element is at least 2 characters example: 0h
+	if len(elem) < 2 {
+		return 0, fmt.Errorf("invalid hardened element: %s", elem)
+	}
+
+	number := strings.TrimSuffix(elem, HardenedPostfix)
+
+	// if the element is unchanged, it means it did not end with correct suffix
+	if number == elem {
+		return 0, fmt.Errorf("invalid hardened element")
+	}
+
+	parsedNum, err := strconv.ParseUint(number, 10, 32)
+	if err != nil {
+		return 0, fmt.Errorf("invalid hardened element")
+	}
+
+	numAsUint32 := uint32(parsedNum)
+
+	// todo check for overflow
+	return hdkeychain.HardenedKeyStart + numAsUint32, nil
+}
+
+func parseNormal(elem string) (uint32, error) {
+	if len(elem) == 0 {
+		return 0, fmt.Errorf("invalid normal element")
+	}
+
+	parsedNum, err := strconv.ParseUint(elem, 10, 32)
+	if err != nil {
+		return 0, fmt.Errorf("invalid normal element")
+	}
+
+	return uint32(parsedNum), nil
+}
+
+func ParsePath(path string) ([]uint32, error) {
+	splitted := strings.Split(path, "/")
+
+	if len(splitted) != ExpectedDerivationDepth {
+		return nil, fmt.Errorf("invalid derivation path length")
+	}
+
+	h1, err := parseHardened(splitted[0])
+	if err != nil {
+		return nil, fmt.Errorf("invalid derivation path element at index 0: %w", err)
+	}
+
+	h2, err := parseHardened(splitted[1])
+	if err != nil {
+		return nil, fmt.Errorf("invalid derivation path element at index 1: %w", err)
+	}
+
+	h3, err := parseHardened(splitted[2])
+	if err != nil {
+		return nil, fmt.Errorf("invalid derivation path element at index 2: %w", err)
+	}
+
+	n4, err := parseNormal(splitted[3])
+	if err != nil {
+		return nil, fmt.Errorf("invalid derivation path element at index 3: %w", err)
+	}
+
+	n5, err := parseNormal(splitted[4])
+	if err != nil {
+		return nil, fmt.Errorf("invalid derivation path element at index 4: %w", err)
+	}
+
+	return []uint32{h1, h2, h3, n4, n5}, nil
+
+}
```

Comment:
```suggestion
func ParsePath(path string) ([]uint32, error) {
	const expectedDepth = 5
	parsers := []func(string) (uint32, error){
		parseHardened,
		parseHardened, 
		parseHardened, 
		parseNormal, 
		parseNormal, 
	}

	splitted := strings.Split(path, "/")
	if len(splitted) != expectedDepth {
		return nil, fmt.Errorf("invalid derivation path length")
	}

	result := make([]uint32, expectedDepth)
	for i, part := range splitted {
		parsed, err := parsers[i](part)
		if err != nil {
			return nil, fmt.Errorf("invalid derivation path element at index %d: %w", i, err)
		}
		result[i] = parsed
	}

	return result, nil
}
```

Reply:
suggestion, you don't have to accept it

Reply:
oh tbh I find this flow a bit convoluted

Current is a bit verbose, but it is super simple. (though it may be one of those things that everybody sees a bit different)

I will leave it as it is.